File: desktop-2/src/lib/browse.ts
/**
 * browse · TS contract for the native Tauri child webview.
 *
 * 2026-06-25 · added so desktop-2 can render real Whop / X / YT / Discord
 * content (sites that refuse iframe embedding via X-Frame-Options: DENY).
 * The Rust counterpart lives at `src-tauri/src/browse.rs`. Bounds are
 * caller-provided so the BrowseOverlay (centered 90vw × 88vh modal) can
 * position the webview inside its layout — no fixed right-rail squeeze.
 *
 * Lifecycle (consumed by BrowseOverlay):
 *   1. React mounts an empty webview-slot div.
 *   2. Effect measures slot.getBoundingClientRect() → calls openBrowsePanel.
 *   3. On window resize OR slot resize → updateBrowsePanelBounds.
 *   4. On overlay close → closeBrowsePanel (destroys child webview).
 *
 * Browser-preview safety: control helpers short-circuit when Tauri is absent.
 * Opening is different: it rejects so the UI renders its honest unavailable
 * state instead of labelling an empty slot as loaded.
 */

export interface BrowsePanelBounds {
  /** Logical x position inside the main window (CSS pixels). */
  x: number;
  /** Logical y position inside the main window (CSS pixels). */
  y: number;
  /** Logical width. */
  width: number;
  /** Logical height. */
  height: number;
}

function isTauriRuntime(): boolean {
  return typeof window !== "undefined" && "__TAURI_INTERNALS__" in window;
}

async function invokeOrNoop<T>(cmd: string, args?: Record<string, unknown>): Promise<T | null> {
  if (!isTauriRuntime()) return null;
  const mod = await import("@tauri-apps/api/core");
  return (await mod.invoke<T>(cmd, args)) ?? null;
}

export interface BrowseOpenOutcome {
  opened_in_app: boolean;
  system_fallback: boolean;
}

/** Open (or re-navigate) the browse panel webview at the given bounds.
 *  Commerce URLs (/checkout, /pay, /billing, /upgrade, /subscribe,
 *  /purchase, /cart) are intercepted by Rust and opened in the system
 *  browser — App Store guideline 3.1.1. */
export async function openBrowsePanel(
  url: string,
  bounds: BrowsePanelBounds,
): Promise<BrowseOpenOutcome> {
  if (!isTauriRuntime()) {
    throw new Error("The in-app browser is available in the installed desktop app.");
  }
  const outcome = await invokeOrNoop<BrowseOpenOutcome>("open_browse_panel", {
    url,
    x: bounds.x,
    y: bounds.y,
    width: bounds.width,
    height: bounds.height,
  });
  if (!outcome) throw new Error("The desktop browser did not return an open result.");
  return outcome;
}

/**
 * G2 · Layer 5 (2026-07-04) · the Rust side now returns explicit
 * "panel missing" errors instead of silently succeeding when the
 * child webview isn't there. For commands where missing == "no-op /
 * already done" (close · update-bounds), suppress the specific
 * error string to preserve the documented "safe to call when panel
 * isn't open" contract. For commands where missing == "operation
 * was meaningful, user should know" (back · forward · reload ·
 * webview_eval), the error bubbles so callers can subscribe to the
 * `browse:crashed` Tauri event and reset their UI.
 */
function isPanelMissing(err: unknown): boolean {
  if (typeof err === "string") return err === "panel missing";
  if (err instanceof Error) return err.message === "panel missing";
  return false;
}

/** Resize / reposition the existing webview without re-navigating.
 *  Silently succeeds if the panel isn't open (documented contract). */
export async function updateBrowsePanelBounds(bounds: BrowsePanelBounds): Promise<void> {
  try {
    await invokeOrNoop<void>("update_browse_panel_bounds", {
      x: bounds.x,
      y: bounds.y,
      width: bounds.width,
      height: bounds.height,
    });
  } catch (err) {
    if (!isPanelMissing(err)) throw err;
    // Panel missing → no-op; the Rust side already emitted
    // `browse:crashed` for any listener that cares.
  }
}

/** Destroy the child webview. Safe to call when panel isn't open. */
export async function closeBrowsePanel(): Promise<void> {
  try {
    await invokeOrNoop<void>("close_browse_panel");
  } catch (err) {
    if (!isPanelMissing(err)) throw err;
  }
}

export async function isBrowsePanelOpen(): Promise<boolean> {
  const result = await invokeOrNoop<boolean>("is_browse_panel_open");
  return result === true;
}

export async function browseBack(): Promise<void> {
  await invokeOrNoop<void>("browse_back");
}

export async function browseForward(): Promise<void> {
  await invokeOrNoop<void>("browse_forward");
}

export async function browseReload(): Promise<void> {
  await invokeOrNoop<void>("browse_reload");
}

/**
 * G2 · Layer 5 (2026-07-04) · canonical crash-detection primitive.
 *
 * Returns:
 *   `"responsive"`   — panel exists AND accepted a no-op eval
 *   `"unresponsive"` — panel exists BUT the wry runtime is dead
 *   `"missing"`      — panel handle gone entirely
 *
 * The Rust command emits the `browse:crashed` Tauri event on the
 * `"unresponsive"` and `"missing"` branches, so a react component
 * that cares about the panel's state should listen for the event
 * AND call this method opportunistically when regaining focus.
 */
export type BrowseHealthStatus = "responsive" | "unresponsive" | "missing";

export interface BrowseHealthReport {
  status: BrowseHealthStatus;
}

export async function browseHealthCheck(): Promise<BrowseHealthReport | null> {
  return invokeOrNoop<BrowseHealthReport>("browse_health_check");
}

/**
 * Tauri event name emitted by browse.rs on every child-webview navigation.
 * Payload is the destination URL as a string. Kept in sync with the Rust
 * constant `URL_CHANGED_EVENT` in src-tauri/src/browse.rs.
 */
export const BROWSE_URL_CHANGED_EVENT = "browse:url-changed";

/**
 * Subscribe to child-webview URL changes. Returns an unsubscribe fn.
 *
 * Silently no-ops in browser preview (no Tauri runtime) — the returned
 * unsubscribe fn is safe to call regardless. Every navigation attempt
 * fires, including commerce URLs that get bounced to the system
 * browser; callers filter what they care about.
 *
 * Callers today:
 *   • BrowseOverlay — keeps `currentUrl` + address bar in sync with
 *     real webview location (was Phase 2 TODO).
 *   • whopBountyCapture — auto-detects Whop bounty URLs so the
 *     agency's Post-to-Whop flow completes without a manual copy-paste.
 */
export async function subscribeBrowseUrlChanges(
  cb: (url: string) => void,
): Promise<() => void> {
  if (!isTauriRuntime()) return () => { /* noop */ };
  try {
    const { listen } = await import("@tauri-apps/api/event");
    const unlisten = await listen<string>(BROWSE_URL_CHANGED_EVENT, (evt) => {
      if (typeof evt.payload === "string" && evt.payload.length > 0) {
        cb(evt.payload);
      }
    });
    return unlisten;
  } catch {
    return () => { /* noop — event API unavailable */ };
  }
}

File: desktop-2/src-tauri/src/browse.rs
// Browse panel · native Tauri child webview.
//
// Ported from `desktop/src-tauri/src/browse.rs` (legacy v0.7.x) to the
// desktop-2 shell on 2026-06-25. Whop / X / YouTube / Discord refuse
// iframe embedding (X-Frame-Options: DENY) so the only way to render
// real content in-app is a native child webview that bypasses iframe
// CSP entirely.
//
// Key difference vs legacy: bounds are CALLER-PROVIDED, not fixed to a
// 560px right rail. desktop-2's BrowseOverlay is a centered 90vw × 88vh
// modal; React measures the webview slot (an empty div) via
// getBoundingClientRect and passes (x, y, width, height) to Rust. On
// window resize or layout change, React calls `update_browse_panel_bounds`
// to reposition. Closing the overlay calls `close_browse_panel` which
// destroys the child webview. This avoids the legacy "squashed workspace"
// problem (560px reservation) Daniel called out 2026-06-25.
//
// Commerce-redirect filter preserved verbatim from legacy — App Store
// guideline 3.1.1 says in-app purchases must NOT happen inside a webview.
// Any nav to /checkout, /pay, /billing, /upgrade, /subscribe, /purchase,
// or /cart is opened in the system browser via tauri-plugin-opener.

use tauri::{
    webview::WebviewBuilder, AppHandle, LogicalPosition, LogicalSize, Manager, WebviewUrl,
};
use tauri_plugin_opener::OpenerExt;
use serde::Serialize;

pub const PANEL_LABEL: &str = "browse_panel";

#[derive(Serialize)]
pub struct BrowseOpenOutcome {
    opened_in_app: bool,
    system_fallback: bool,
}

const BLOCKED_PATH_FRAGMENTS: &[&str] = &[
    "/checkout",
    "/pay",
    "/billing",
    "/upgrade",
    "/subscribe",
    "/purchase",
    "/cart",
];

fn is_commerce_url(url: &tauri::Url) -> bool {
    let path = url.path().to_lowercase();
    BLOCKED_PATH_FRAGMENTS.iter().any(|frag| path.contains(frag))
}

/// Open or re-navigate the browse panel webview.
///
/// `x`, `y`, `width`, `height` are LOGICAL window coordinates (device
/// pixels divided by scale factor). React measures the webview slot
/// element via getBoundingClientRect and passes the values directly.
#[tauri::command]
pub async fn open_browse_panel(
    app: AppHandle,
    url: String,
    x: f64,
    y: f64,
    width: f64,
    height: f64,
) -> Result<BrowseOpenOutcome, String> {
    let parsed_url: tauri::Url = url
        .parse()
        .map_err(|e| format!("invalid url: {e}"))?;

    // Commerce filter — open in system browser instead of in-app.
    if is_commerce_url(&parsed_url) {
        app.opener()
            .open_url(parsed_url.to_string(), None::<&str>)
            .map_err(|e| format!("system browser fallback failed: {e}"))?;
        return Ok(BrowseOpenOutcome {
            opened_in_app: false,
            system_fallback: true,
        });
    }

    // Re-navigate if already open.
    if let Some(existing) = app.get_webview(PANEL_LABEL) {
        existing
            .navigate(parsed_url)
            .map_err(|e| format!("navigate failed: {e}"))?;
        // Update position in case React layout changed since last open.
        let _ = existing.set_position(LogicalPosition::new(x.max(0.0), y.max(0.0)));
        let _ = existing.set_size(LogicalSize::new(width.max(120.0), height.max(120.0)));
        return Ok(BrowseOpenOutcome {
            opened_in_app: true,
            system_fallback: false,
        });
    }

    // Spawn new child webview with commerce-redirect navigation filter.
    let main = app
        .get_window("main")
        .ok_or_else(|| "main window not found".to_string())?;

    // 2026-06-30 · pin the webview profile to a known directory inside
    // the app's data store so cookies + localStorage survive overlay
    // close + relaunch. Platform behaviour:
    //   · Windows WebView2 · data_directory IS the cookie jar location ·
    //     without this call WebView2 picks a temp-scoped folder that
    //     gets wiped on app restart for some user profiles.
    //   · macOS WKWebView · effectively a no-op · WKWebsiteDataStore.
    //     default() is already persistent and tied to the app bundle ID
    //     (app.liquidclips.desktop). The call is preserved for cross-
    //     platform consistency + Windows correctness.
    //   · Linux WebKitGTK · respected if the build links a recent enough
    //     wry; harmless if not.
    let profile_dir = app
        .path()
        .app_data_dir()
        .map_err(|e| format!("app_data_dir unavailable: {e}"))?
        .join("webview_profile");

    let app_for_filter = app.clone();
    let builder = WebviewBuilder::new(PANEL_LABEL, WebviewUrl::External(parsed_url))
        .data_directory(profile_dir)
        .on_navigation(move |nav_url| {
            if is_commerce_url(nav_url) {
                let target = nav_url.to_string();
                let app = app_for_filter.clone();
                tauri::async_runtime::spawn(async move {
                    let _ = app.opener().open_url(target, None::<&str>);
                });
                return false;
            }
            true
        });

    main.add_child(
        builder,
        LogicalPosition::new(x.max(0.0), y.max(0.0)),
        LogicalSize::new(width.max(120.0), height.max(120.0)),
    )
    .map_err(|e| format!("add_child failed: {e}"))?;
    Ok(BrowseOpenOutcome {
        opened_in_app: true,
        system_fallback: false,
    })
}

/// Resize / reposition the existing webview without re-navigating.
/// Called by React on window-resize / overlay-layout-change.
#[tauri::command]
pub async fn update_browse_panel_bounds(
    app: AppHandle,
    x: f64,
    y: f64,
    width: f64,
    height: f64,
) -> Result<(), String> {
    if let Some(wv) = app.get_webview(PANEL_LABEL) {
        wv.set_position(LogicalPosition::new(x.max(0.0), y.max(0.0)))
            .map_err(|e| format!("set_position failed: {e}"))?;
        wv.set_size(LogicalSize::new(width.max(120.0), height.max(120.0)))
            .map_err(|e| format!("set_size failed: {e}"))?;
    }
    Ok(())
}

/// Destroy the child webview. Called on overlay close + on app shutdown.
#[tauri::command]
pub async fn close_browse_panel(app: AppHandle) -> Result<(), String> {
    if let Some(wv) = app.get_webview(PANEL_LABEL) {
        wv.close().map_err(|e| format!("close failed: {e}"))?;
    }
    Ok(())
}

#[tauri::command]
pub async fn is_browse_panel_open(app: AppHandle) -> bool {
    app.get_webview(PANEL_LABEL).is_some()
}

#[tauri::command]
pub async fn browse_back(app: AppHandle) -> Result<(), String> {
    if let Some(wv) = app.get_webview(PANEL_LABEL) {
        wv.eval("window.history.back()")
            .map_err(|e| format!("back eval failed: {e}"))?;
    }
    Ok(())
}

#[tauri::command]
pub async fn browse_forward(app: AppHandle) -> Result<(), String> {
    if let Some(wv) = app.get_webview(PANEL_LABEL) {
        wv.eval("window.history.forward()")
            .map_err(|e| format!("forward eval failed: {e}"))?;
    }
    Ok(())
}

#[tauri::command]
pub async fn browse_reload(app: AppHandle) -> Result<(), String> {
    if let Some(wv) = app.get_webview(PANEL_LABEL) {
        wv.eval("window.location.reload()")
            .map_err(|e| format!("reload eval failed: {e}"))?;
    }
    Ok(())
}

/// F6 (Layer 3 · reliability sprint) — evaluate arbitrary JavaScript inside
/// the persistent browse-panel webview. Fire-and-forget: `wv.eval` on Tauri
/// v2 does not return the JS return value. Callers that need a result MUST
/// wrap the script so it emits a Tauri event (or calls `window.__TAURI_IPC`)
/// with the payload. The F6 Gmail-compose driver uses this pattern to poll
/// for compose-ready + send-confirmation + captcha states.
///
/// Errors:
///   - "browse panel not open"          · caller must open_browse_panel first
///   - "webview eval failed: <detail>"  · the WebView refused the script
///                                       (rare — usually the caller passed
///                                       malformed JS or the webview died)
#[tauri::command]
pub async fn webview_eval(app: AppHandle, script: String) -> Result<(), String> {
    let wv = app
        .get_webview(PANEL_LABEL)
        .ok_or_else(|| "browse panel not open".to_string())?;
    wv.eval(&script)
        .map_err(|e| format!("webview eval failed: {e}"))?;
    Ok(())
}
